### PR TITLE
Support batching in `Composite` operators

### PR DESCRIPTION
**Pull request by AlbertMitjans**
This PR does the following:
- `Composite`:
  - Check that all operands have the same `batch_size` and update the `_batch_size` attribute at instantiation.
- `Prod`:
  - Use `matmul` instead of `dot` to support batching (see attached image).
  - Change the `matrix` method to allow broadcasting in multiple operands (see example below).
- `SymbolicOp`:
  - Move `batch_size` property to this class and remove it from its children.

image

### Broadcasting with multiple operands:
Now we can do:
```pycon
>>> op1 = qml.RX(np.array([1, 2, 3]), 0)
>>> op2 = qml.RY(np.array([4, 5, 6]), 2)
>>> prod_op = qml.prod(op1, op2)
>>> prod_op.matrix().shape
(3, 4, 4)
```
In master, we obtain the following:
```pycon
>>> op1 = qml.RX(np.array([1, 2, 3]), 0)
>>> op2 = qml.RY(np.array([4, 5, 6]), 2)
>>> prod_op = qml.prod(op1, op2)
>>> prod_op.matrix().shape
(9, 4, 4)
```
Because `np.kron` doesn't allow broadcasting, thus `kron` is applied to all axes.

### CODE REVIEW

**Comment by AlbertMitjans** [comment]
[sc-31244]

File: pennylane/ops/op_math/composite.py
Comment (timmysilv) [inline pennylane/ops/op_math/composite.py:68]: I would prefer to call this `_set_batch_size` since that's ultimately what it's hoping to do. Also, is it in its own method just for cleanliness? not picky either way, just curious

File: pennylane/ops/op_math/prod.py
Comment (timmysilv) [inline pennylane/ops/op_math/prod.py:267]: might some `op.batch_size` values ever be `0` rather than `None`, and if so would that break things?

File: tests/ops/op_math/test_composite.py
Comment (timmysilv) [inline tests/ops/op_math/test_composite.py:137]: the `RY` would ultimately be converted into a batch_size-3 equivalent if this was a real composite operation, right? If so, whose responsibility is it to ensure it, and should the CompositeOp implementation know whose responsibility it is/if it's upheld later? eg. prod takes care of the matrix work in this PR, but what if it didn't?

File: pennylane/ops/op_math/composite.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/composite.py:68]: Changed!

File: pennylane/ops/op_math/prod.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/prod.py:267]: I believe they are never 0 nor 1, even though having `batch_size = 1` for a non-broadcasted operator would make more sense for me.

@dwierichs why was this decision taken?

File: tests/ops/op_math/test_composite.py
Comment (AlbertMitjans) [reply tests/ops/op_math/test_composite.py:137]: To avoid changing classes dynamically, `RY` is not converted into a batch_size-3 equivalent:
```pycon
>>> base = qml.RX(np.array([1.2, 2.3, 3.4]), 0)
>>> qml.prod(base, qml.RY(1, 0), qml.RZ(np.array([1, 2, 3]), wires=2))
RX(array([1.2, 2.3, 3.4]), wires=[0]) @ RY(1, wires=[0]) @ RZ(array([1, 2, 3]), wires=[2])
```

Right now, the `matrix` method is responsible to ensure batching is available. If I'm correct, the whole batching pipeline doesn't have a way to determine if batching is available for an operator or not. Thus if you use batching with an operator that doesn't support it, either you get a weird error, or you get the wrong result.

Comment (AlbertMitjans) [reply tests/ops/op_math/test_composite.py:137]: But again, I'm sure @dwierichs has a better answer

File: pennylane/ops/op_math/prod.py
Comment (timmysilv) [reply pennylane/ops/op_math/prod.py:267]: right, 1 is better than 0 😛 I won't block this PR with this concern, but if `Operator` has the default value of 1 for batch_size, I think the checking of batch_size throughout the codebase would be a lot simpler. This looks good to me, but I'd like to make that change after this PR (unless @dwierichs has concerns we're missing, that is)

**Review by timmysilv — APPROVED** [review]
looks super! the `yield` stuff was kinda fancy so I'll miss it lol but this is definitely easier to follow and great to have. I'd prefer we wait until David weighs in on the open conversations before merging, but from where I stand this is good to go

File: pennylane/ops/op_math/prod.py
Comment (dwierichs) [inline pennylane/ops/op_math/prod.py:249]: Just a comment: I think this logic `qml.matrix(H) if isinstance(H, qml.Hamiltonian) else H.matrix()` exists a lot in the codebase. Maybe we want to create a functionality for this at some point that doesn't require writing this out every time?

File: pennylane/ops/op_math/prod.py
Comment (dwierichs) [inline pennylane/ops/op_math/prod.py:251]: Could we get away with memorizing the information `batched` and using it later on, instead of copying the matrix? :thinking:

File: pennylane/ops/op_math/prod.py
Comment (dwierichs) [reply pennylane/ops/op_math/prod.py:267]: `batch_size=None` indicates a non-broadcasted operation, which to me is different from a broadcasted operation with dimension `1`. I see that one could argue that they are the same, because they contain the same amount of information. However, without any broadcasting, an additional axis of size `1` seems very unintuitive to me. Also, `qml.RX(np.array([0.4]), wires=0)` and `qml.RX(0.4, wires=0)` would yield the same batch dimension if we set `1` as the default that currently is `None`. 
Finally, all features that do not support a broadcasting axis yet would become unsupported even without broadcasting if we replace `None` by `1` and consider it as "dummy broadcasting"

File: pennylane/ops/op_math/prod.py
Comment (dwierichs) [reply pennylane/ops/op_math/prod.py:251]: And another thought: Do we need to `stack` here, if we anyways manually iterate over the stacked axis later on (in the `reduce(math.kron...` line)?

File: tests/ops/op_math/test_composite.py
Comment (dwierichs) [inline tests/ops/op_math/test_composite.py:135]:
```suggestion
        """Test that the batch_size is correct when some but not all operands are batched."""
```

File: tests/ops/op_math/test_composite.py
Comment (dwierichs) [reply tests/ops/op_math/test_composite.py:137]: >  If so, whose responsibility is it to ensure it,

An operation that claims to support batching should, well, support batching. In particular that means that the `matrix` method of the operation needs to know what to do in case broadcasting/batching is used.
For `CompositeOp`s, this in particular means that it has to support a variety of combinations (all input ops batched, some batched, none batched...).

When executing a quantum tape, for example, no conversion is necessary, because devices that support broadcasting will use that matrix multiplication can be broadcasted itself.

Note that there is `qml.ops.qubit.attributes.supports_broadcasting` which contains all operations that support broadcasting. I think this PR should add the updated operations to that attribute :)

Comment (dwierichs) [reply tests/ops/op_math/test_composite.py:137]: > Thus if you use batching with an operator that doesn't support it, either you get a weird error, or you get the wrong result.

That's right. Typically you will get an error because of dimension mismatches. This is not special for broadcasting though: If we want to ensure operations are not instantiated with ill-formatted inputs, the operations have to check themselves that the input is okay.

**Review by dwierichs — COMMENTED** [review]
Very nice PR adding batching support, thanks a lot! :100:

There is the somewhat artificial (I think) question whether we also want to allow broadcasting in the following sense:
```python
qml.prod(np.array([qml.RX(0.4, 0), qml.RZ(0.3, 0), qml.RY(0.21, 0)]), qml.PauliZ(0))
```
I think we don't want to go there because of the huge space of combinations that could appear. Nonetheless, as broadcasting support usually means that the operation inputs can have a leading batch dimension (as opposed to getting inputs that are batched themselves), we should be very clear in the documentation about this. Maybe even add a warning that the above shenanigans are not possible.

Regarding the default value of `batch_size` I left a comment, I hope this helps! :)

Finally, I'm not sure whether we want to add the composite ops to `qml.ops.qubit.attributes.supports_broadcasting`. I mean, they now do support broadcasting, but as mentioned above in a different way than, say, `RX`.

File: pennylane/ops/op_math/prod.py
Comment (timmysilv) [reply pennylane/ops/op_math/prod.py:267]: I was under the impression that broadcasting support is done on a case-by-case basis, essentially via `op.matrix()` or something like that. For this reason, I was presuming that an operator could simply ignore the concept of batching if the batch_size is 1.

> `qml.RX(np.array([0.4]), wires=0)` and `qml.RX(0.4, wires=0)` would yield the same batch dimension

Is this an implementation detail, or is there actual value gained from this distinction?

File: pennylane/ops/op_math/prod.py
Comment (timmysilv) [reply pennylane/ops/op_math/prod.py:249]: I think `qml.matrix` _is_ that functionality, but I'm guessing people avoid just doing `qml.matrix` because of the overhead of it being wrapped as an op_transform?

Comment (josh146) [reply pennylane/ops/op_math/prod.py:249]: This I think may also be solved by the new op arithmetic, in that `qml.matrix()` will no longer be 'opinionated', but will simply always try and compute the matrix.

Which... has other downfalls. The matrix computation could take forever if you are not careful.

Comment (timmysilv) [reply pennylane/ops/op_math/prod.py:249]: this makes me wonder - I have more than once in the past taken advantage of the fact that `qml.matrix()` works with tapes (because of the `tape_transform` side effect defined as `_matrix()`). Perhaps that functionality should be explicitly separated, and the user should _know_ they're looking for the matrix of a tape/series of operations. With that separated out, we would simply use `op.matrix()` (or `tape.matrix()`) when relevant.

File: tests/ops/op_math/test_composite.py
Comment (AlbertMitjans) [reply tests/ops/op_math/test_composite.py:137]: Added `Prod` and `Sum` into `qml.ops.qubit.attributes.supports_broadcasting`

**Comment by AlbertMitjans** [comment]
> There is the somewhat artificial (I think) question whether we also want to allow broadcasting in the following sense:
```python
qml.prod(np.array([qml.RX(0.4, 0), qml.RZ(0.3, 0), qml.RY(0.21, 0)]), qml.PauliZ(0))
```

Yeah I'm also against allowing this for the moment. We can think of adding it in the future, if it can be super useful for users.

> we should be very clear in the documentation about this

Where is batching explained? Should we add it in the docstring of `Prod` and `Sum`?

> Maybe even add a warning that the above shenanigans are not possible.

Right now using an array in `prod` or `sum` raises the following error:
```pycon
AttributeError: 'tensor' object has no attribute 'wires'
```

I believe this is enough to tell the users that this is not supported 😆 . But we can add an specific error if preferred.

File: pennylane/ops/op_math/prod.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/prod.py]: Nice catch!! Created a list containing bools that specify if the matrix is batched or not, and use it later to re-use the matrix.

File: pennylane/ops/op_math/prod.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/prod.py]: Honestly I don't know why we are doing this...

I just profiled both functions and we don't have that much overhead:
```pycon
>>> import pennylane as qml
>>> import timeit
>>> prod = qml.prod(qml.PauliX(0), qml.PauliY(1))
>>> %timeit prod.matrix()
22.1 µs ± 90.3 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
>>> %timeit qml.matrix(prod)
24.4 µs ± 227 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

I changed this line to use `qml.matrix`.

Comment (dwierichs) [reply pennylane/ops/op_math/prod.py]: I feel like @albi3ro could have an additional insight and opinion here as well, having thought a lot about operator arithmetic.

File: pennylane/ops/op_math/prod.py
Comment (dwierichs) [reply pennylane/ops/op_math/prod.py]: Nice! :100:

**Comment by dwierichs** [comment]
Thanks for the updates @AlbertMitjans ! 

> Where is batching explained? Should we add it in the docstring of Prod and Sum?

There is information on using broadcasting/batching in QNodes [here](https://docs.pennylane.ai/en/stable/introduction/circuits.html?highlight=broadcasting#parameter-broadcasting-in-qnodes) and more technical information [here](https://docs.pennylane.ai/en/stable/code/api/pennylane.operation.Operator.html?highlight=Operator#parameter-broadcasting) One could consider adding a drop-down menu on broadcasting in the `Prod` and/or `Sum` operation documentation, if there is sufficiently much information we want to put there.

> I believe this is enough to tell the users that this is not supported

Ah, yes, sorry, I was not clear with what I meant: One could add a ..warning:: or ..admonition:: block in the docstring :)

**Comment by AlbertMitjans** [comment]
@dwierichs added a note in both `Sum` and `Prod` docstrings saying that batching of operators is not supported

File: pennylane/ops/op_math/prod.py
Comment (albi3ro) [reply pennylane/ops/op_math/prod.py]: So Hamiltonian is a special case where it just never got `Hamiltonian.matrix` defined. I think this idea was "this could potentially be expensive, so we don't want to define it on the operation.". I do have an issue open to move `qml.utils.sparse_hamiltonian` to the `Hamiltonian.sparse_matrix` method. This has just been a low-priority code organization improvement.

I think we should use the `Operator.matrix` method internally, especially in cases like this. 

If we were planning on maintaining `Hamiltonian` in the long term, I would say we should add `Hamiltonian.matrix`, but I don't think it's necessarily worth the effort now.

Comment (AlbertMitjans) [reply pennylane/ops/op_math/prod.py]: Mmm ok. Maybe this `2us` difference can get noticeable when having tons of operands...

I'll revert it to using `op.matrix()`

**Review by dwierichs — APPROVED** [review]
Looks great :100:
Thanks for the docstring comments regarding broadcasting!

File: pennylane/ops/op_math/prod.py
Comment (dwierichs) [reply pennylane/ops/op_math/prod.py]: Maybe this is just a subjective thing, but I actually like the distinction. From my perspective, passing a broadcasted axis of size `1` that is not _intended_ to be a broadcasting axis points to a user code error.

Comment (dwierichs) [reply pennylane/ops/op_math/prod.py]: But maybe this is something we should discuss in a broader context :)

File: pennylane/ops/op_math/prod.py
Comment (dwierichs) [reply pennylane/ops/op_math/prod.py]: Should this conversation be closed? :)

**Pull request merged**
